src/Messages/OWSMessageSender.m

```diff
@@ -697,26 +677,12 @@ - (void)groupSend:(NSArray<SignalRecipient *> *)recipients
     [completionFuture catchDo:^(id failure) {
         // failure from toc_thenAll yeilds an array of failed Futures, rather than the future's failure.
         if ([failure isKindOfClass:[NSArray class]]) {
-            NSArray *errors = (NSArray *)failure;
-            for (TOCFuture *failedFuture in errors) {
-                if (!failedFuture.hasFailed) {
-                    // If at least one send succeeded, don't show message as failed.
-                    // Else user will tap-to-resend to all recipients, including those that already received the
-                    // message.
-                    return successHandler();
-                }
-            }
-
-            // At this point, all recipients must have failed.
-            // But we have all this verbose type checking because TOCFuture doesn't expose type information.
-            id lastError = errors.lastObject;
-            if ([lastError isKindOfClass:[TOCFuture class]]) {
-                TOCFuture *failedFuture = (TOCFuture *)lastError;
-                if (failedFuture.hasFailed) {
-                    id failureResult = failedFuture.forceGetFailure;
+            NSArray *groupSendFutures = (NSArray *)failure;
+            for (TOCFuture *groupSendFuture in groupSendFutures) {
```

Comment:
As I understand this change, we're now going to mark a group send as failed, if a single recipient fails to receive it. 

That has an obvious up side. The downside is that we'll confuse the situation where multiple recipients do get it but, e.g. the one unregistered group member, does not get it.

Is there a way to address this without going down the design-rabbit-hole of per-recipient delivery status? Just brainstorming... like wrapping the error and displaying it in the retry copy or footer status.

Even if not, I think what you have is a more conservative (better in this case) approach.